File: js_server/src/routes/oauth.ts
import * as Router from 'koa-router'
import { MyRouter } from '../typeDeclear'
import axios from 'axios'
import { callback_url, config, oauth_url } from '../config'
import * as dayjs from 'dayjs'

export const router = new Router() as MyRouter

// oauth/callback
router.get('/callback', async ctx => {
    let code = ctx.query.code

    if (!code) {
        ctx.redirect(oauth_url)
        return
    }
    try {
        let r = await axios.post('https://bgm.tv/oauth/access_token',
            {
                'grant_type': 'authorization_code',
                'client_id': config.appID,
                'client_secret': config.appSecret,
                'code': code,
                'redirect_uri': callback_url
            })
        if (r.data.error) {
            ctx.status = 400
            ctx.body = r.data
            return
        } else {
            r.data.auth_time = dayjs(r.headers['Date']).unix()

            await ctx.db.collection('token').findOne({ '_id': r.data['user_id'] })
            await ctx.db.collection('token').updateOne({ '_id': r.data['user_id'] },
                { '$set': r.data },
                { upsert: true })
            return ctx.render('post_to_extension.html', { 'data': JSON.stringify(r.data), })
        }
    } catch (e) {
        console.log(e)
        ctx.redirect(oauth_url)
    }
})

router.get('/result', async ctx => {
    ctx.body = dayjs('Sat, 13 Oct 2018 10:06:18 GMT').unix()
    // await ctx.render('post_to_extension.html', { data: { hello: 'world' } })
})
router.get('/redirect', async ctx => {
    ctx.redirect(oauth_url)
})

File: server/views/oauth.py
import json

import aiohttp.client_exceptions
import aiohttp_jinja2
from aiohttp import web
from dateutil import parser as date_parser

from server.app_types import WebRequest
from server.config import oauth_url, APP_ID, APP_SECRET, callback_url


async def get_token(request: WebRequest, ):
    code = request.query.get('code', None)
    if not code:
        raise web.HTTPFound(location=oauth_url)
    async with request.app.client_session.post(
        'https://bgm.tv/oauth/access_token',
        data={'grant_type'   : 'authorization_code',
              'client_id'    : APP_ID,
              'client_secret': APP_SECRET,
              'code'         : code,
              'redirect_uri' : callback_url}) as resp:
        try:
            r = await resp.json()
        except aiohttp.client_exceptions.ContentTypeError:
            raise web.HTTPFound(oauth_url)
    if 'error' in r:
        return web.json_response(r, status=400)
    r['auth_time'] = int(date_parser.parse(resp.headers['Date']).timestamp())
    async with request.app.client_session.get(
        'https://api.bgm.tv/user/{}'.format(r['user_id'])
    ) as resp:
        try:
            user_info = await resp.json()
        except aiohttp.client_exceptions.ContentTypeError:
            raise web.HTTPFound(oauth_url)

    session = await request.session.new_session()
    session.update(r)
    session.update(user_info)
    session.login = True
    return aiohttp_jinja2.render_template('post_to_extension.html',
                                          request,
                                          {'data': json.dumps(r)})


async def refresh_auth_token(request: WebRequest, ):
    data = await request.json()
    refresh_token = data.get('refresh_token', None)
    user_id = data.get('user_id', None)
    if not (refresh_token and user_id):
        return web.HTTPBadRequest()
    data = {'grant_type'   : 'refresh_token',
            'client_id'    : APP_ID,
            'client_secret': APP_SECRET,
            'refresh_token': refresh_token,
            'redirect_uri' : callback_url}
    async with request.app.client_session.post(
        'https://bgm.tv/oauth/access_token', json=data
    ) as resp:
        try:
            r = await resp.json()
        except aiohttp.client_exceptions.ContentTypeError:
            return web.json_response({}, status=504)
    if 'error' in r:
        return web.json_response(r)
    r['auth_time'] = int(date_parser.parse(resp.headers['Date']).timestamp())

    async with request.app.client_session.get(
        'https://api.bgm.tv/user/{}'.format(r['user_id'])
    ) as resp:
        try:
            user_info = await resp.json()
        except aiohttp.client_exceptions.ContentTypeError:
            raise web.HTTPFound(oauth_url)

    if request.session.login:
        request.session.update(r)
    else:
        s = await request.session.new_session()
        s.update(r)
    request.session.update(user_info)
    return web.json_response(r)
